[17기 이동근]ADD: categoryproduct View

gledong12 opened the pull request:
>wecode PR message template 및 체크 리스트 입니다. 
아래 사항들을 전부 작성/체크 하시고 PR 해주세요!

## 수정 사항 간략한 한줄 요약
(여기에 수정 사항에 대한 간략한 한줄 요약/제목 작성해 주세요.)
 - main 페이지의 전체보기 View 작성완료(프론트와 연결 성공)
 - 각 카테고리별로 데이터 보내기 위한 View작성 완료(프론트와 연결은 확인하지 못함)
		 
## 수정 사항들 자세한 내용
(여기에 수정 사항에 대한 자세한 내용을 작성해 주세요.)

## 기타 질문 및 특이 사항
(궁금한 사항들, 하면서 느낀 점들 공유해주실 것이 있으면 자유롭게 작성해 주세요.)
- 현재 전체보기 View와, 각 카테고리별로 데이터를 보기위한 View를 따로 나누었는데, 
쿼리문을 사용해서 한개의 View로 합치고 싶은데 이를 오류처리를 하자니 너무 편법인거 같고해서 혹시 두 개의 View를 한 개로 합칠 수 있을까요? 

## 체크 리스트 (아래 사항들이 전부 체크되어야만 merge가 됩니다!)
- [ ] 필요한 test들을 완료하였고 기능이 제대로 실행되는지 확인 하였습니다.
- [x] Wecode의 코드 스타일 가이드에 맞추어 코드를 작성 하였습니다.
- [ ] 제가 의도한 파일들과 수정 사항들만 커밋이 된 것을 확인 하였습니다.
- [ ] 본 수정 사항들을 팀원들과 사전에 상의하였고 팀원들 모두 해당 PR에 대하여 알고 있습니다.

Commit:
ADD: categoryproduct View

Commit:
MODIFY: 기업명 추가

jamessoun93 commented:
수정해주세요 :)
<img width="594" alt="Screen Shot 2021-02-22 at 10 44 13 AM" src="https://user-images.githubusercontent.com/50141312/108647009-e9515680-74fa-11eb-8e10-7c6be1e1ab47.png">

jamessoun93 commented on WeGotDiz/settings.py:
```diff
@@ -80,7 +80,7 @@
 # https://docs.djangoproject.com/en/3.1/ref/settings/#databases
 
-DATABASES = my_settings.DATABASES
+DATABASES = my_settings.DATABASE
```
이 부분은 팀원들과 상의 후 변경하신건가요? 아니라면 꼭 다같이 변경할 수 있도록 해주세요!

gledong12 replied:
my_setting 수정했습니다.!!

jamessoun93 commented on WeGotDiz/urls.py:
```diff
@@ -1,5 +1,6 @@
+from django.urls import path, include
 
 urlpatterns = [
+        path('product', include('product.urls'))
```
4칸이면 충분합니다!
```suggestion
    path('product', include('product.urls'))
```

gledong12 replied:
수정했습니다.!!

jamessoun93 commented on user/models.py:
```diff
@@ -3,7 +3,7 @@
     fullname    = models.CharField(max_length=20)
     image       = models.URLField(max_length=2000)
-    email       = models.EmailField(max_length=200, unique=True)
+    email       = models.EmailField(max_length=200)
```
email 중복이 가능한가요?!

gledong12 replied:
수정했습니다.!!

jamessoun93 commented on product/models.py:
```diff
@@ -1,4 +1,5 @@
 from django.db         import models
+from django.db.models.deletion import CASCADE
```
불필요합니다!! 그냥 없이 사용했더니 안되셔서 추가하신걸까요??

gledong12 replied:
???? 머죠저게... 자동으로 추가된 것 같습니다... 주의하겠습니다.!!

jamessoun93 commented on product/urls.py:
```diff
@@ -0,0 +1,8 @@
+from django.urls import path, include
+from .views      import CategoryView, CategoryDetailView
+
+urlpatterns = [
+        path('/category', CategoryView.as_view()),
+        path('/category/<int:Category_id>', CategoryDetailView.as_view())
+
+]
```
- whitespace 관리 신경써주세요 😁
- 그리고 path param 활용 잘하셨습니다 👍
```suggestion
urlpatterns = [
    path('/category', CategoryView.as_view()),
    path('/category/<int:Category_id>', CategoryDetailView.as_view())
]
```

gledong12 replied:
수정 완료했습니다.!! 감사합니다.!

jamessoun93 commented on product/views.py:
```diff
@@ -1,3 +1,62 @@
-from django.shortcuts import render
+import json
+import bcrypt
+import jwt
 
-# Create your views here.
+from django.http              import HttpResponse, JsonResponse
+from django.views             import View
+
+from datetime                 import date, datetime
+from .models                  import Category, Product, Reward, Collection, LikeUser, ProductContent
```
깔끔하네요!! 🥳
하지만 datetime 은 파이썬 내장 모듈이므로 위로 올려주시는게 맞겠네요!

gledong12 replied:
넵 수정 완료했습니다.!!

jamessoun93 commented on product/views.py:
```diff
@@ -1,3 +1,62 @@
+class CategoryView(View):
+    def get(self, request):
+        try:
+            product = Product.objects.all()
+            product_list = [{
+                'category_image'   : product.category_set.all()[0].image,
+                'title'            : product.title,
+                'goal_amout'       : product.goal_amount,
+                'toal_amount'      : product.total_amount,
+                'achieved_rate'    : product.achieved_rate,
+                'total_supporters' : product.total_supporters,
+                'closing_date'     : product.closing_date,
+                'thumbnail'        : product.thumbnail_url,
+                'category'         : product.category_set.all()[0].name,
+                'id'               : product.id,
+                'maker_info_name'  : product.maker_info.name
+                } for product in product]
+            return JsonResponse({"MESSAGE" : "SUCCESS", "data" : {'product'  : product_list}}, status=200)
```
- list comprehension 적용 완벽합니다 👍
- 가독성을 위해 빈 줄 추가해주시면 더욱 더 깔끔해지겠네요!
- `Product.objects.all()` 의 결과물은 QuerySet list 이므로 변수명도 복수형으로 지어주세요!
- `all()[0]` 을 `first()` 라는 메소드로 대체할 수도 있으니 참고바랍니다 :)
```suggestion
            product = Product.objects.all()
            
            product_list = [{
                'category_image'   : product.category_set.all()[0].image,
                'title'            : product.title,
                'goal_amout'       : product.goal_amount,
                'toal_amount'      : product.total_amount,
                'achieved_rate'    : product.achieved_rate,
                'total_supporters' : product.total_supporters,
                'closing_date'     : product.closing_date,
                'thumbnail'        : product.thumbnail_url,
                'category'         : product.category_set.all()[0].name,
                'id'               : product.id,
                'maker_info_name'  : product.maker_info.name
                } for product in product]
                
            return JsonResponse({"MESSAGE" : "SUCCESS", "data" : {'product'  : product_list}}, status=200)
```

gledong12 replied:
오호 꿀팁 감사합니다.!!! 또 하나 배워 갑니다.!!

jamessoun93 commented on product/views.py:
```diff
@@ -1,3 +1,62 @@
+class CategoryView(View):
+    def get(self, request):
+        try:
+            product = Product.objects.all()
+            product_list = [{
+                'category_image'   : product.category_set.all()[0].image,
+                'title'            : product.title,
+                'goal_amout'       : product.goal_amount,
+                'toal_amount'      : product.total_amount,
+                'achieved_rate'    : product.achieved_rate,
+                'total_supporters' : product.total_supporters,
+                'closing_date'     : product.closing_date,
+                'thumbnail'        : product.thumbnail_url,
+                'category'         : product.category_set.all()[0].name,
+                'id'               : product.id,
+                'maker_info_name'  : product.maker_info.name
+                } for product in product]
+            return JsonResponse({"MESSAGE" : "SUCCESS", "data" : {'product'  : product_list}}, status=200)
```
return 해주시는 json body 를 펼쳐보면 아래처럼 뭔가 언밸런스한 느낌을 받는데요! key 값을 소문자로 통일하고 불필요한 space 를 없애서 조금 더 깔끔하게 만들 수 있을 것 같습니다.
```json
{
    "MESSAGE" : "SUCCESS", 
    "data" : {
        'product'  : product_list
    }
}
```
```suggestion
            product = Product.objects.all()
            product_list = [{
                'category_image'   : product.category_set.all()[0].image,
                'title'            : product.title,
                'goal_amout'       : product.goal_amount,
                'toal_amount'      : product.total_amount,
                'achieved_rate'    : product.achieved_rate,
                'total_supporters' : product.total_supporters,
                'closing_date'     : product.closing_date,
                'thumbnail'        : product.thumbnail_url,
                'category'         : product.category_set.all()[0].name,
                'id'               : product.id,
                'maker_info_name'  : product.maker_info.name
                } for product in product]
            return JsonResponse({"message" : "SUCCESS", "data" : {'product' : product_list}}, status=200)
```

gledong12 replied:
수정완료 했습니다!!!

jamessoun93 commented on product/views.py:
```diff
@@ -1,3 +1,62 @@
+                'maker_info_name'  : product.maker_info.name
+                } for product in product]
+            return JsonResponse({"MESSAGE" : "SUCCESS", "data" : {'product'  : product_list}}, status=200)
+        except KeyError:
+            return JsonResponse({"MESSAGE" : "INVALIED_KEY"}, status=400)
+
```
그리고 여기서 KeyError 를 잡아주신 이유를 설명해주시겠어요? comment 로 답글 달아주세요 👍

gledong12 replied:
아 CategoryView와 CategoryDetailView를 한개의 View로 합치는 도중 오류처리 해줄려다가 남겨 뒀던것 같습니다. CategoryView 에는 굳이 keyError 처리가 필요없는데 괜히 해주었네요 수정하겠습니다 !!

jamessoun93 commented on product/views.py:
```diff
@@ -1,3 +1,62 @@
+from .models                  import Category, Product, Reward, Collection, LikeUser, ProductContent
+
+
+class CategoryView(View):
```
클래스명은 카테고리 목록을 뿌려주는 view class 라는것을 한눈에 알수있게 지어주시는게 좋습니다. 저는 아래쪽으로 내려가서 detail view 를 확인해서 알고있지만 다른 파일에서 이 클래스를 import 할때 헷갈릴 수 있습니다~

jamessoun93 commented on product/views.py:
```diff
@@ -1,3 +1,62 @@
+            return JsonResponse({"MESSAGE" : "INVALIED_KEY"}, status=400)
+
+
+class CategoryDetailView(View):
+    def get(self, request, Category_id):
+        try:
+            category_id = request.GET.get('category', None)
+            data = Category.objects.get(id=category_id)
+            relatd_product = data.product.all()
+        
+            product_list = [{
+                'category_image'   : data.image,
+                'title'            : product.title,
+                'goal_amout'       : product.goal_amount,
+                'toal_amount'      : product.total_amount,
+                'achieved_rate'    : product.achieved_rate,
+                'total_supporters' : product.total_supporters,
+                'closing_date'     : product.closing_date,
+                'thumbnail'        : product.thumbnail_url,
+                'category'         : data.name,
+                'id'               : product.id,
+                'maker_info_name'  : product.maker_info.name
+                } for product in relatd_product]
+            return JsonResponse({"MESSAGE" : "SUCCESS", "data" : {'product'  : product_list}}, status=200)
+        except KeyError:
+            return JsonResponse({"MESSAGE" : "INVALIED_KEY"}, status=400)
+
```
view 잘 작성하셨는데 테스트를 제대로 안해보고 올리신 흔적이 몇군데 있어서, 테스트 과정에서 고칠 수 있는 부분 고치시고 다시 push 해주세요 동근님 🧐

jamessoun93 reviewed (changes requested):
수고하셨습니다 동근님 🤩
아래 내용과 코드에 남겨드린 리뷰 반영하여 push 해주세요! 
- csv 파일들과 uploader 파일은 git 에 올려 공유하는 파일들이 아니므로 없애주시고 gitignore 에도 추가해주세요!
- 그리고 view 를 작성하면 꼭 httpie 혹은 postman 으로 테스트를 하신 후 push 해주세요!

jamessoun93 commented:
> 현재 전체보기 View와, 각 카테고리별로 데이터를 보기위한 View를 따로 나누었는데,
쿼리문을 사용해서 한개의 View로 합치고 싶은데 이를 오류처리를 하자니 너무 편법인거 같고해서 혹시 두 개의 View를 한 개로 합칠 수 있을까요?

좋은 질문입니다~! 두개의 View 는 각각 완전히 다른 정보를 반환해주기 때문에 합칠 이유가 없습니다. 합칠 수는 있지만, 합친다면 조건문에 의해 굉장히 가독성이 안좋은 코드가 나오게 될거에요!

Commit:
MODIFY: ProductlistVIw로 수정, 두 개의 View를 한개로 수정 완료

gledong12 commented:
CategoryView 와 CategoryDetatil VIew -> ProductlistView로 수정 완료 햇습니다.
필터 기능 추가 하면서 수정하겠습니다.!

gledong12 committed:
Delete csv directory

push된 csv 파일 삭제하겠습니다.!!
제가 .gitignore에는 .csv라고 추가가 되어있는데, 혹시 csv파일을 따로 csv폴더에 모아놓게 되면 .gitignore에 추가해도 push가 되는 건가요?

gledong12 commented:
지금 Category 관련해서 조건을 부여 후 로직에서 product_list를 한번만 사용할려고 합니다. 
우선, Product는 filter를 사용하려고 하고, Category는 Category_id=None인 경우와, 아닌 경우 로 나눠 각각리스트를 작성해서 JsonResponse를 각각 보낼 수 있도록 하려고 합니다.

먼가 될 것 같으면서도 잘 안되서 그런데 혹시 None을 처리하는 방법에 대해서 힌트 같은게 있을까요?

Commit:
MODIFY: productlistView 작성완료

gledong12 commented:
어제 conflict난거 수정하느라 filter작업하던 내용이 그대로 push가 된 것 같습니다. query랑 url에서 category_id를 받아서 실행시키는것 모두 확인했습니다. 

conflict를 해결하는 과저에서 git clone을 다시 받아서 작업해둔 내용을 복사 후 push하는 것이 더울 좋을 것 같아서 commit이 많아 보이는데 수정은 urls.py와 Views.py 만 수정됬습니다.!!


승현님이 가르쳐주신 Q를 사용하는 법 공부했는데 단순히 카테고리를 보여주는 View 에서는 굳이 사용하지 않아도 될 것 같아서  filter구현할때 사용해 보도록 하겠습니다.!!

Commit:
MODIFY: 필터링 제외한 나머지 부분push

gledong12 commented:
멘토님 제가 머지가 되야지 다른 분들 머지가 된다고 해서  지금 올린거 머지 해주실 수 있나요? 필터링은 제가 다른 branch 파서 구현하겠습니다.

Commit:
MODIFY: 카테고리 데이터 객체화
